### PR TITLE
LUCENE-8232: Separate out PendingDeletes from ReadersAndUpdates

### PR DESCRIPTION
**Pull request by s1monw**
Today ReadersAndUpdates is tightly coupled with IW and all the
handling of pending deletes. This change decouples IW and pending
deletes from ReadersAndUpdates and allows expert users to
customize how deletes are handled. This is useful or up to a
certain extend mandatory to work with soft-deletes to allow
merge policies to make the right decisions.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/index/DefaultPendingDeletes.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+import org.apache.lucene.codecs.LiveDocsFormat;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.TrackingDirectoryWrapper;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.IOUtils;
+import org.apache.lucene.util.MutableBits;
+
+/**
+ * This class handles accounting and applying pending deletes for live segment readers
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/index/DefaultPendingDeletes.java]: Maybe say "hard" or "real" deletes here?

File: lucene/core/src/java/org/apache/lucene/index/IndexWriter.java [hunk 2]
```diff
@@ -604,7 +604,10 @@ public synchronized void release(ReadersAndUpdates rld, boolean assertInfoLive)
         if (!poolReaders && rld.refCount() == 1 && readerMap.containsKey(rld.info)) {
           // This is the last ref to this RLD, and we're not
           // pooling, so remove it:
-          if (rld.writeLiveDocs(directory)) {
+          boolean changed = rld.writeLiveDocs(directory);
+          changed |= rld.writeFieldUpdates(directory, globalFieldNumberMap, bufferedUpdatesStream.getCompletedDelGen(), infoStream);
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/index/IndexWriter.java]: Whoa this looks like a pre-existing bug, that we failed to checkpoint if the only thing that changed was new DV updates written?

Comment (s1monw) [reply lucene/core/src/java/org/apache/lucene/index/IndexWriter.java]: yeah that might  in-fact be true here

**Comment by s1monw** [comment]
this has been merged

**Pull request closed without merging**